packages/ckeditor5-ui/src/editorui/editorui.ts

```diff
@@ -393,18 +398,43 @@ export default abstract class EditorUI extends /* #__PURE__ */ ObservableMixin()
 		menuBarView.fillFromConfig( normalizedMenuBarConfig, this.componentFactory, this._extraMenuBarElements );
 
 		this.editor.keystrokes.set( 'Esc', ( data, cancel ) => {
-			if ( menuBarViewElement.contains( this.editor.ui.focusTracker.focusedElement ) ) {
+			if ( !menuBarViewElement.contains( this.editor.ui.focusTracker.focusedElement ) ) {
+				return;
+			}
+
+			// Bring focus back to where it came from before focusing the toolbar:
+			// If it came from outside the engine view (e.g. source editing), move it there.
+			if ( this._lastFocusedForeignElement ) {
+				this._lastFocusedForeignElement.focus();
+				this._lastFocusedForeignElement = null;
+			}
+			// Else just focus the view editing.
+			else {
 				this.editor.editing.view.focus();
-				cancel();
 			}
+
+			cancel();
 		} );
 
 		this.editor.keystrokes.set( 'Alt+F9', ( data, cancel ) => {
-			if ( !menuBarViewElement.contains( this.editor.ui.focusTracker.focusedElement ) ) {
-				menuBarView.isFocusBorderEnabled = true;
-				menuBarView!.focus();
-				cancel();
+			// If menu bar is already focused do nothing.
+			if ( menuBarViewElement.contains( this.editor.ui.focusTracker.focusedElement ) ) {
+				return;
+			}
+
+			const focusedElement = this.focusTracker.focusedElement as HTMLElement;
+
+			// If focus moved out of a DOM element that does not belong to the editing view (e.g. source editing).
+			if (
+				Array.from( this._editableElementsMap.values() ).includes( focusedElement ) &&
+				!Array.from( this.editor.editing.view.domRoots.values() ).includes( focusedElement )
+			) {
+				this._lastFocusedForeignElement = focusedElement;
```

Comment:
Could we extract this to a common (for alt+10 and alt+f9 listeners) private method like `_updateLastFocusedElement()`? It looks like this bit repeats in both listeners. WDYT?

Reply:
Okay, I can do that